CaloCluster/src/CaloClusterFast_module.cc

```diff
@@ -1,322 +1,197 @@
-                                        CaloHitCollection&              recoCrystalHits)
- {
-
-	const CaloDigiCollection& caloDigis(*caloDigisHandle);
-
-	mu2e::GeomHandle<mu2e::Calorimeter> ch;
-	const Calorimeter* cal = ch.get();
-	int nro = cal->caloInfo().getInt("nSiPMPerCrystal");
-
-	unsigned offsetT0_ = unsigned(blindTime_/digiSampling_);
-	unsigned nBinTime  = unsigned (mbtime_ - blindTime_ + endTimeBuffer_) / digiSampling_;
-	unsigned winOffsetT0_ = windowPeak_+offsetT0_ ;
-
-	if (hitList_.size() < nBinTime ) hitList_ = std::vector<std::list<FastHit>>(nBinTime,std::list<FastHit>());
-	else for (auto& hl : hitList_) hl.clear();
-
-	seeds_.clear();
-
-
-	for (const auto& caloDigi : caloDigis)
-      	{
-           if (caloDigi.SiPMID() % nro) continue;
-
-           double t0    = caloDigi.t0();
-           int    roId  = caloDigi.SiPMID();
-           int    crId  = roId / nro; //cal->caloIDMapper().crystalIDFromSiPMID(roId); use short form for efficiency
-           //double adc2MeV  = calorimeterCalibrations->ADC2MeV(roId);
-           const auto& waveform = caloDigi.waveform();
-
-           auto it   = waveform.begin();
-           auto tail = waveform.begin();
-           unsigned countdown(window_ - 1);
-           unsigned nCount(0);
-           deque_.clear();
-	  
-           for (; it != waveform.end(); ++it)
-           {
-                while (!deque_.empty() && *it > deque_.back()) deque_.pop_back();
-                deque_.push_back(*it);
-
-                if (countdown > 0) countdown--;
-                else
-                {
-                   if (deque_.front()> minAmp_ && deque_.front()== *std::prev(it,windowPeak_) && *std::prev(it,windowPeak_) != *std::prev(it,windowPeak_-1))
-                   {
-                       int index = int(t0/digiSampling_) + nCount - winOffsetT0_;
-		      
-                       hitList_[index].push_back(FastHit(crId,index,deque_.front()));
-                       if (deque_.front()> minSeedAmp_) seeds_.emplace_back(&(hitList_[index].back()));
-                   }
-
-                   if (*tail == deque_.front()) deque_.pop_front();
-                   ++tail;
-                }
-                ++nCount;
-           } 
+  //----------------------------------------------------------------------------------------------------------
+  void CaloClusterFast::makeClusters(CaloClusterCollection& caloClusters, const art::Handle<CaloHitCollection>& caloHitsHandle)
+  {
+      const Calorimeter& cal = *(GeomHandle<Calorimeter>());
+      const CaloHitCollection& caloHits(*caloHitsHandle);
+      if (caloHits.empty()) return;
+
+      std::vector<int> hits(caloHits.size());
+      std::iota(hits.begin(),hits.end(),0);
+      
+      auto functorTime = [&caloHits](int a, int b) {return caloHits[a].time() < caloHits[b].time();};
+      std::sort(hits.begin(),hits.end(),functorTime);
+      for (auto val : hits) if (caloHits[val].energyDep() < EnoiseCut_) val = -1;
```

Comment:
could you make this cut 1 step earlier, ie only fill the hits index array with hits above thershold.  This would reduce the number of downstream operations.

Reply:
That would also eliminate the clumsy logic having to interpret -1 below.

Reply:
Yes. In practice, almost all hits will pass the cut but I have done the change anyway